### PR TITLE
[JENKINS-73760] Updates fail due to invalid JSON from HTTP Update Center

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-73760](https://issues.jenkins.io/browse/JENKINS-73760). Partial revert of #6116, which deprecated/removed useful functionality (the ability to migrate old update sites to new ones). #2996 should have used this subsystem but did not, resulting in [JENKINS-73760](https://issues.jenkins.io/browse/JENKINS-73760). This PR restores the subsystem and correctly applies it to #2996 to complete the migration from http://updates.jenkins-ci.org to https://updates.jenkins.io which was begun in 2017.

### Testing done

New unit tests.

### Proposed changelog entries

Migrate from http://updates.jenkins-ci.org to https://updates.jenkins.io when the initial installation version was 2.76 or older.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (basil) [inline core/src/main/java/hudson/model/UpdateSite.java:542]: Reverting this hunk of #6116, since this functionality is useful and should not be deprecated.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (basil) [inline core/src/main/java/hudson/model/UpdateSite.java:541]: This was just wrong. This method was introduced in 1.357.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (basil) [inline core/src/main/java/hudson/model/UpdateSite.java:545]: Reverting this hunk of #6116, since this functionality is useful and should not be deprecated.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (basil) [inline core/src/main/java/hudson/model/UpdateSite.java:551]: Adapting the pre-#6116 code to handle the migration from http://updates.jenkins-ci.org/ to https://updates.jenkins.io/ rather than the migration from http://hudson-ci.org to http://updates.jenkins-ci.org/

File: test/src/test/java/hudson/model/UpdateCenterMigrationTest.java
Comment (basil) [inline test/src/test/java/hudson/model/UpdateCenterMigrationTest.java:25]: `hudson.model.UpdateCenter.xml` taken from a stock 2.76 installation.

File: test/src/test/java/hudson/model/UpdateSiteTest.java
Comment (basil) [inline test/src/test/java/hudson/model/UpdateSiteTest.java:221]: Adapting the pre-#6116 code to handle the migration from http://updates.jenkins-ci.org/ to https://updates.jenkins.io/ rather than the migration from http://hudson-ci.org to http://updates.jenkins-ci.org/

**Review by timja — APPROVED** [review]
Thanks!

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
